Assistant: Improve managing language models

georgestagg opened the pull request:
This PR makes a few changes in an attempt to improve managing LLM provider models.

The manage models quickpick menu has been improved:
 * Copilot's BYOK providers, particularly ones that Assistant does not currently support (e.g. Grok) are no longer shown.
 * Only providers that the user has signed in with are shown
 * The `managementCommand` property has been added to Assistant providers, so that the cog icon works.
 * These changes address #9283.

A new configuration setting `positron.assistant.filterModels` has been added. With this users (and in the future admins) can set glob filters. Models will only be shown if they match one of the given globs.
 * This addresses #9788.

A `positron.assistant.preferredModel` setting has been added. With this, when the Positron window first loads it should default to loading the model given in that setting. Here we use partial matching on the ID or model names.
* This partially addresses #9388, in that IIUC it matches the behaviour requested in the [final comment](https://github.com/posit-dev/positron/issues/9388#issuecomment-3282375011) of that thread.

A `positron.assistant.defaultModels` settings has been added. With this, users can specify the default model for a particular provider, which we will pre-select in the model selector if the user hasn't specified a preferred model instead.
* This addresses the remaining parts of https://github.com/posit-dev/positron/issues/9388 based on https://github.com/posit-dev/positron/pull/9799#discussion_r2408682852.
* We have set the default model for Anthropic to `Claude Sonnet 4` and Bedrock to `Claude 4 Sonnet Bedrock`

---

### Release Notes

#### New Features

- Assistant: added setting `positron.assistant.filterModels` to configure available models (#9788)
- Assistant: added setting `positron.assistant.preferredModel` to set the preferred model to pre-select in the model selector (#9388)
- Assistant: added setting `positron.assistant.defaultModels` to set the default model per provider (#9388)

#### Bug Fixes

- Assistant: non-functional providers removed from "Manage models..." quickpick (#9283)


### QA Notes

#### Manage models quickpick
1. First, ensure you have multiple providers configured and signed in. I used Anthropic, AWS Bedrock, and GitHub Copilot Chat.
2. In the Assistant panel, invoke the model picker: <img width="403" height="173" alt="Screenshot 2025-10-06 at 15 42 22" src="https://github.com/user-attachments/assets/2cc33831-37c3-4f46-a1fc-887f32d93311" />
3. Ensure only useful entries are shown: <img width="620" height="196" alt="Screenshot 2025-10-06 at 15 42 27" src="https://github.com/user-attachments/assets/d644d1ea-f198-4bdb-bc3c-752c0fe8e20a" />
4. Select a _signed in_ model, ensure a model list appears:
<img width="617" height="297" alt="Screenshot 2025-10-06 at 15 42 39" src="https://github.com/user-attachments/assets/9fcef6a6-f623-4349-8529-27f1e6713bb3" />

5. Do the same but select a signed out model, ensure the model configuration modal appears.
6. Do the same thing with a _signed in_ model but click the cog on the far right, the model configuration modal should again appear.

#### Model filtering

1. In your user settings, set the following two filters:
```
"positron.assistant.filterModels": ["Claude*", "GPT-5"],
```
2. Restart Positron or Reload the window.
3. Ensure that only Claude family models or GPT-5 can be used: 
<img width="366" height="238" alt="Screenshot 2025-10-06 at 15 49 04" src="https://github.com/user-attachments/assets/3e7ac3ed-b9bd-41a8-b8a4-ca7b2bf0aca3" />

4. This should also work with IDs, useful for e.g. OpenRouter:
```
"positron.assistant.filterModels": ["**/google/gemini-*"],
```
<img width="431" height="389" alt="Screenshot 2025-10-06 at 16 07 43" src="https://github.com/user-attachments/assets/c54e9720-ab3b-429b-a33d-4485dbc454fd" />

5. Finally, this should also restrict the manage models quickpick:
<img width="604" height="476" alt="Screenshot 2025-10-06 at 16 08 20" src="https://github.com/user-attachments/assets/e1c45c9e-ca16-46c6-9f51-0ca3140dd954" />

#### Setting preferred model

1. In your user settings, remove the `positron.assistant.filterModels` setting and set the following:
```
"positron.assistant.preferredModel": "GPT-5",
```
2. In the Assistant chat panel, switch to another provider and model, e.g. AWS Bedrock and Claude 4 Sonnet.
3. Create a new chat, ensure selected model does not change.
4. Reload the window, ensure that on reload the preferred model set in the configuration is first selected. In this case, it should switch to GPT-5 from the Copilot provider on reload.

#### Setting default model per provider

1. Sign into multiple providers, at least Anthropic and AWS Bedrock.
2. Try switching between them, confirm that the defaults we want by default are selected.
3. Set the following settings,
```
  "positron.assistant.defaultModels": { "anthropic-api": "Opus", "amazon-bedrock": "Opus" },
```
4. Reload the window, confirm that switching between the models now selects Opus models by default.

Note 1: Upstream has code to remember your last selection, so after the reload window you have to change providers to see the default kick in.

Note 2: Sometimes Assistant gets stuck in a state showing "Pick Model..." in the selector. Switching providers does nothing to clear that, but selecting a model first, and then switching providers, gets things working again. I don't know if this is related to this commit, or the changes in #9799.

georgestagg committed:
Assistant: Improve "Manage Models" interaction

georgestagg committed:
Assistant: Add model filtering configuration

georgestagg committed:
Assistant: Override persisted model by configuration

georgestagg committed:
Prefer model exact match when searching

georgestagg commented:
Looks like I've broken something, will be busy over the new few days but will debug failing tests when I can.

georgestagg commented:
Sure, feel free to commit straight to this branch if you figure it out.

sharon-wang committed:
keep Echo language model in package.json

sharon-wang committed:
Filter "manage models" action to only providers that are signed in

- required an update to the display name of AWS Bedrock to Amazon Bedrock in our UI (to match the name in package.json)

sharon-wang commented:
Pushed a couple changes to

- hopefully fix the test error by adding the Echo language model back d521546b098da65544a138f44a1d733612545922
- filter the "manage models" quickpick to only show vendors/providers that the user has already logged into 4e65121dcec972eee3d9b5cab30cbc3806952bfc

sharon-wang committed:
fix failing language model tests

Run test locally via `./scripts/test.sh --grep "LanguageModels"`

georgestagg commented on src/vs/workbench/contrib/chat/common/languageModels.ts:
```diff
@@ -405,6 +413,13 @@ export class LanguageModelsService implements ILanguageModelsService {
 				}
 
+				// --- Start Positron ---
+				// Don't include BYOK model providers from copilot-chat extension
+				if (extension.description.id === 'GitHub.copilot-chat' && extension.value && Array.isArray(extension.value)) {
+					extension.value = extension.value.filter(v => v.vendor === 'copilot');
+				}
```
This should filter out only the providers that GitHub Copilot Chat includes as BYOK providers. In Positron, we own that functionality via the model configuration modal and our own provider SDK (and vercel) integrations, so we can remove them. We keep just the base Copilot Chat provider and its models.

Saying that, if the model selector quickpick now only shows logged in models we can remove this patch entirely since they won't show up anyway.

georgestagg commented on src/vs/workbench/contrib/chat/browser/actions/manageModelsActions.ts:
```diff
@@ -76,6 +85,12 @@ export class ManageModelsAction extends Action2 {
+				// --- Start Positron ---
+				// Show model configuration if no models are registered for this vendor yet
+				if (models.length === 0 && selectedItem.managementCommand === 'positron-assistant.configureModels') {
+					commandService.executeCommand(selectedItem.managementCommand);
```
I think hitting this code path will be extremely rare now. If you get as far as signing in, we can be pretty sure there are models available since we test one just to sign in.

I would suggest, in the interests of minimising the number of upstream patches, just removing this now.

georgestagg replied:
Removed in [e8824b9](https://github.com/posit-dev/positron/pull/9799/commits/e8824b93bcc5e3b8297c4b454ddb787bc9efe2db).

georgestagg commented on extensions/positron-assistant/package.json:
```diff
@@ -287,31 +287,51 @@
+              "type": "string"
+            }
+          },
+          "positron.assistant.preferredModel": {
+            "type": "string",
+            "default": "",
+            "markdownDescription": "%configuration.preferredModel.description%"
```
I think these should be different settings. My intention for `preferredModel` is that it would be a global setting for a single model and switch to the named model, even if from a different provider.

> This would allow us to set a default model for e.g. Anthropic to Claude 4 Sonnet, which should provide a better initial experience than e.g. Haiku.

Do we think people will want to override this choice? Upstream already has a concept of a default model for a provider, so we could set that metadata property to `true` in the Assistant extension for named models that we choose. Then users who care to default elsewhere can use `preferredModel` as-is.

I will make a PR based on this one to demonstrate.

georgestagg committed:
Match vendor filter by ID

georgestagg committed:
Remove no-longer needed upstream patches

georgestagg committed:
Assistant: Add and apply default models setting

sharon-wang committed:
add settings examples and logging for filtered models

sharon-wang committed:
don't log cache control data part skips

causes unnecessary debug noise in the logs

sharon-wang committed:
use the preferred model when switching providers

sharon-wang committed:
switch default anthropic model to claude sonnet 4

this is the model we've done the most testing with and seems to work well, even though claude sonnet 4.5 is newer

sharon-wang committed:
add examples for default model setting and note precedence of preferred model over default model

sharon-wang committed:
update comments re: determining the selected model

timtmok commented on extensions/positron-assistant/src/anthropic.ts:
```diff
@@ -334,10 +348,9 @@ export class AnthropicLanguageModel implements positron.ai.LanguageModelChatProv
 					maxOutputTokens: maxOutputTokens,
 					capabilities: this.capabilities,
-					isDefault: isFirst,
+					isDefault: this.isDefaultUserModel(model.id, model.display_name),
```
I'm not sure if the `defaultModels` setting works as intended. If I set the default to `Claude Sonnet 4.5`, it does set it as default but `Claude Sonnet 4` is also set as default. I end up with 4 selected and not 4.5.

timtmok replied:
I've changed it so that there will never be more than one default model.

timtmok committed:
Fix default Anthropic model

Ensures there cannot be more than one default model

timtmok committed:
Fix current provider when filtered out

Sets a new current provider if it is filtered out

timtmok commented on src/vs/workbench/contrib/chat/common/languageModels.ts:
```diff
@@ -616,14 +624,35 @@ export class LanguageModelsService implements ILanguageModelsService {
+					// Get and apply LLM allow filters from configuration.
+					const config = this._configurationService.getValue<{ filterModels: string[] }>('positron.assistant');
+					this._logService.trace('[LM] Applying model filters:', config.filterModels);
+					if (config.filterModels.length > 0 && !config.filterModels.some(pattern =>
```
I've added some eventing around the `filterModels` setting so that the provider selector updates.
